pyiceberg/table/__init__.py

```diff
@@ -350,6 +357,241 @@ class RemovePropertiesUpdate(TableUpdate):
+
+    if context.last_added_schema_id is not None and context.last_added_schema_id == update.schema_id:
+        context.updates.append(SetCurrentSchemaUpdate(schema_id=-1))
+    else:
+        context.updates.append(update)
+
+    return TableMetadataUtil.parse_obj(updated_metadata_data)
+
+
+@apply_table_update.register(AddSnapshotUpdate)
+def _(update: AddSnapshotUpdate, base_metadata: TableMetadata, context: TableMetadataUpdateContext) -> TableMetadata:
+    if len(base_metadata.schemas) == 0:
+        raise ValueError("Attempting to add a snapshot before a schema is added")
+
+    if len(base_metadata.partition_specs) == 0:
+        raise ValueError("Attempting to add a snapshot before a partition spec is added")
+
+    if len(base_metadata.sort_orders) == 0:
+        raise ValueError("Attempting to add a snapshot before a sort order is added")
+
+    if base_metadata.snapshots_by_id.get(update.snapshot.snapshot_id) is not None:
+        raise ValueError(f"Snapshot with id {update.snapshot.snapshot_id} already exists")
+
+    if (
+        base_metadata.format_version == 2
+        and update.snapshot.sequence_number is not None
+        and update.snapshot.sequence_number <= base_metadata.last_sequence_number
+        and update.snapshot.parent_snapshot_id is not None
+    ):
+        raise ValueError(
+            f"Cannot add snapshot with sequence number {update.snapshot.sequence_number} "
+            f"older than last sequence number {base_metadata.last_sequence_number}"
+        )
+
+    updated_metadata_data = copy(base_metadata.model_dump())
+    updated_metadata_data["last-updated-ms"] = update.snapshot.timestamp_ms
+    updated_metadata_data["last-sequence-number"] = update.snapshot.sequence_number
+    updated_metadata_data["snapshots"].append(update.snapshot.model_dump())
+    context.updates.append(update)
+    return TableMetadataUtil.parse_obj(updated_metadata_data)
+
+
+@apply_table_update.register(SetSnapshotRefUpdate)
+def _(update: SetSnapshotRefUpdate, base_metadata: TableMetadata, context: TableMetadataUpdateContext) -> TableMetadata:
+    if update.type is None:
+        raise ValueError("Snapshot ref type must be set")
+
+    if update.min_snapshots_to_keep is not None and update.type == SnapshotRefType.TAG:
+        raise ValueError("Cannot set min snapshots to keep for branch refs")
+
+    if update.min_snapshots_to_keep is not None and update.min_snapshots_to_keep <= 0:
+        raise ValueError("Minimum snapshots to keep must be >= 0")
+
+    if update.max_snapshot_age_ms is not None and update.type == SnapshotRefType.TAG:
+        raise ValueError("Tags do not support setting maxSnapshotAgeMs")
+
+    if update.max_snapshot_age_ms is not None and update.max_snapshot_age_ms <= 0:
+        raise ValueError("Max snapshot age must be > 0 ms")
+
+    if update.max_ref_age_ms is not None and update.max_ref_age_ms <= 0:
+        raise ValueError("Max ref age must be > 0 ms")
```

Comment:
```suggestion
    if update.type is None:
        raise ValueError("Snapshot ref type must be set")
    elif update.min_snapshots_to_keep is not None and update.type == SnapshotRefType.TAG:
        raise ValueError("Cannot set min snapshots to keep for branch refs")
    elif update.min_snapshots_to_keep is not None and update.min_snapshots_to_keep <= 0:
        raise ValueError("Minimum snapshots to keep must be >= 0")
    elif update.max_snapshot_age_ms is not None and update.type == SnapshotRefType.TAG:
        raise ValueError("Tags do not support setting maxSnapshotAgeMs")
    elif update.max_snapshot_age_ms is not None and update.max_snapshot_age_ms <= 0:
        raise ValueError("Max snapshot age must be > 0 ms")
    elif update.max_ref_age_ms is not None and update.max_ref_age_ms <= 0:
        raise ValueError("Max ref age must be > 0 ms")
```

Some more observations:

- The first check `update.type is None` should never pass because `type` is a required field.
- We could also enforce the `> 0` constraints using Pydantic: https://docs.pydantic.dev/latest/concepts/types/#composing-types-via-annotated

Reply:
Thanks for the suggestion! I moved all the validations to `SnapshotRef` model in `refs.py`. (Pydantic is great!)

There is one weird thing I observed:
I first tried to construct an optional int field in the way suggested by [pydantic numeric constraints](https://docs.pydantic.dev/latest/concepts/fields/#numeric-constraints):
```python
min_snapshots_to_keep: Optional[Annotated[int, Field(alias="min-snapshots-to-keep", default=None, gt=0)]]
```
However this will make `min_snapshots_to_keep` a required field since I get validation error when constructing SnapshotRef instance without this field

Later I found that defining the type like this:
```python
 min_snapshots_to_keep: Annotated[Optional[int], Field(alias="min-snapshots-to-keep", default=None, gt=0)]
```
could make everything work as expected.

I've added some tests in `test_refs.py` for this change.